### PR TITLE
Spec: Refactoring: Merge "definitions" and "declarations".

### PR DESCRIPTION
**Pull request by sjrd**
Previously, abstract members were syntactically separate from concrete members, and referred to "declarations", as opposed to "definitions".

In this commit, all of them become "definitions". There are abstract definitions and concrete definitions. They share their syntactic productions.

This will allow to more easily introduce more kinds of definitions that can be abstract or concrete in deeper productions of the grammar, such as extension methods.

Tentatively assigning to @smarter

### CODE REVIEW

File: docs/_spec/04-basic-definitions.md [hunk 1]
```diff
@@ -20,20 +16,18 @@ Def         ::=  PatVarDef
...
-
-The scope of a name introduced by a declaration or definition is the whole statement sequence containing the binding.
+The scope of a name introduced by a definition is the whole statement sequence containing the definition.
 However, there is a restriction on forward references in blocks:
 In a statement sequence ´s_1 ... s_n´ making up a block, if a simple name in ´s_i´ refers to an entity defined by ´s_j´ where ´j \geq i´, then for all ´s_k´ between and including ´s_i´ and ´s_j´,
 
 - ´s_k´ cannot be a variable definition.
 - If ´s_k´ is a value definition, it must be lazy.
 
+Moreover, in a block, all definitions must be concrete, and opaque type alias definitions are not allowed.
```
Comment (smarter) [inline docs/_spec/04-basic-definitions.md:29]: It's not very useful, but abstract type definitions are allowed in blocks.

File: docs/_spec/04-basic-definitions.md [hunk 2]
```diff
@@ -457,7 +452,7 @@ Variance annotations indicate how instances of parameterized types vary with res
 A ‘+’ variance indicates a covariant dependency, a ‘-’ variance indicates a contravariant dependency, and a missing variance indication indicates an invariant dependency.
 
 A variance annotation constrains the way the annotated type variable may appear in the type or class which binds the type parameter.
-In a type definition `type ´T´[´\mathit{tps}\,´] = ´S´`, or a type declaration `type ´T´[´\mathit{tps}\,´] >: ´L´ <: ´U´` type parameters labeled ‘+’ must only appear in covariant position whereas type parameters labeled ‘-’ must only appear in contravariant position.
+In a type definition `type ´T´[´\mathit{tps}\,´] = ´S´`, `type ´T´[´\mathit{tps}\,´] >: ´L´ <: ´U´` or `type ´T´[´\mathit{tps}\,´] >: ´L´ <: ´U´ = ´S´`, type parameters labeled ‘+’ must only appear in covariant position whereas type parameters labeled ‘-’ must only appear in contravariant position.
```
Comment (smarter) [inline docs/_spec/04-basic-definitions.md:455]:
```suggestion
In a type definition `type ´T´[´\mathit{tps}\,´] = ´S´`, `type ´T´[´\mathit{tps}\,´] >: ´L´ <: ´U´` or `opaque type ´T´[´\mathit{tps}\,´] >: ´L´ <: ´U´ = ´S´`, type parameters labeled ‘+’ must only appear in covariant position whereas type parameters labeled ‘-’ must only appear in contravariant position.
```

File: docs/_spec/04-basic-definitions.md [hunk 3]
```diff
@@ -562,15 +555,15 @@ ParamType          ::=  Type
...
 ```
 
-A _method declaration_ has the form `def ´f\,\mathit{psig}´: ´T´`, where ´f´ is the method's name, ´\mathit{psig}´ is its parameter signature and ´T´ is its result type.
-A _method definition_ `def ´f\,\mathit{psig}´: ´T´ = ´e´` also includes a _method body_ ´e´, i.e. an expression which defines the method's result.
+An _abstract method definition_ has the form `def ´f\,\mathit{psig}´: ´T´`, where ´f´ is the method's name, ´\mathit{psig}´ is its parameter signature and ´T´ is its result type.
+A _concrete method definition_ `def ´f\,\mathit{psig}´: ´T´ = ´e´` also includes a _method body_ ´e´, i.e. an expression which defines the method's result.
 A parameter signature consists of an optional type parameter clause `[´\mathit{tps}\,´]`, followed by zero or more value parameter clauses `(´\mathit{ps}_1´)...(´\mathit{ps}_n´)`.
-Such a declaration or definition introduces a value with a (possibly polymorphic) method type whose parameter types and result type are as given.
+Such a definition introduces a method with a type (if there is no type or term parameter clause) or methodic type whose parameter types and result type are as given (otherwise).
```
Comment (smarter) [inline docs/_spec/04-basic-definitions.md:561]: This sentence is hard to parse, are we distinguishing "method with a type" versus "method with a methodic type"? Could the former be called "method with a proper type" to avoid confusion?

**Review by smarter — APPROVED** [review]
Otherwise LGTM, nice refactoring!

File: docs/_spec/04-basic-definitions.md (hunk 3 above)
Comment (sjrd) [reply docs/_spec/04-basic-definitions.md]: Is it better now?

**Review by smarter — APPROVED** [review]
